sql/colflow: TestVectorizedFlowShutdown failed

[(sql/colflow).TestVectorizedFlowShutdown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1833620&tab=buildLog) on [master@8f793a0d2a20ec75d65a2bd86120d15b16632dde](https://github.com/cockroachdb/cockroach/commits/8f793a0d2a20ec75d65a2bd86120d15b16632dde):

```
=== RUN   TestVectorizedFlowShutdown
--- FAIL: TestVectorizedFlowShutdown (1.43s)
=== RUN   TestVectorizedFlowShutdown/shutdownScenario=ConsumerClosed#05
W200327 15:53:34.490308 474 sql/colflow/colrpc/outbox.go:162  [streamID=0] Outbox Send header error, distributed query will fail: EOF
    --- FAIL: TestVectorizedFlowShutdown/shutdownScenario=ConsumerClosed#05 (0.03s)
        vectorized_flow_shutdown_test.go:364: 
            	Error Trace:	vectorized_flow_shutdown_test.go:364
            	Error:      	Should be true
            	Test:       	TestVectorizedFlowShutdown/shutdownScenario=ConsumerClosed#05
            	Messages:   	outbox with ID 0 did not call Close on closers

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestVectorizedFlowShutdown PKG=./pkg/sql/colflow TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestVectorizedFlowShutdown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colflow).TestVectorizedFlowShutdown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1836219&tab=buildLog) on [master@92495ffb2182671b3338a8c6cd2734cb6de5d392](https://github.com/cockroachdb/cockroach/commits/92495ffb2182671b3338a8c6cd2734cb6de5d392):

```
=== RUN   TestVectorizedFlowShutdown
--- FAIL: TestVectorizedFlowShutdown (1.16s)
=== RUN   TestVectorizedFlowShutdown/shutdownScenario=ConsumerClosed#03
W200329 05:23:33.713524 392 sql/colflow/colrpc/outbox.go:162  [streamID=5] Outbox Send header error, distributed query will fail: EOF
    --- FAIL: TestVectorizedFlowShutdown/shutdownScenario=ConsumerClosed#03 (0.01s)
        vectorized_flow_shutdown_test.go:364: 
            	Error Trace:	vectorized_flow_shutdown_test.go:364
            	Error:      	Should be true
            	Test:       	TestVectorizedFlowShutdown/shutdownScenario=ConsumerClosed#03
            	Messages:   	outbox with ID 5 did not call Close on closers

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestVectorizedFlowShutdown PKG=./pkg/sql/colflow TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestVectorizedFlowShutdown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colflow/colrpc/outbox.go
+++ b/pkg/sql/colflow/colrpc/outbox.go
@@ -103,6 +103,16 @@ func NewOutbox(
 	return o, nil
 }
 
+func (o *Outbox) close(ctx context.Context) {
+	for _, closer := range o.closers {
+		if err := closer.IdempotentClose(ctx); err != nil {
+			if log.V(1) {
+				log.Infof(ctx, "error closing Closer: %v", err)
+			}
+		}
+	}
+}
+
 // Run starts an outbox by connecting to the provided node and pushing
 // coldata.Batches over the stream after sending a header with the provided flow
 // and stream ID. Note that an extra goroutine is spawned so that Recv may be
@@ -131,39 +141,47 @@ func (o *Outbox) Run(
 ) {
 	o.runnerCtx = ctx
 	ctx = logtags.AddTag(ctx, "streamID", streamID)
-
 	log.VEventf(ctx, 2, "Outbox Dialing %s", nodeID)
-	conn, err := dialer.Dial(ctx, nodeID, rpc.DefaultClass)
-	if err != nil {
-		log.Warningf(
-			ctx,
-			"Outbox Dial connection error, distributed query will fail: %+v",
-			err,
-		)
-		return
-	}
 
-	client := execinfrapb.NewDistSQLClient(conn)
-	stream, err := client.FlowStream(ctx)
-	if err != nil {
-		log.Warningf(
-			ctx,
-			"Outbox FlowStream connection error, distributed query will fail: %+v",
-			err,
-		)
-		return
-	}
+	var stream execinfrapb.DistSQL_FlowStreamClient
+	if err := func() error {
+		conn, err := dialer.Dial(ctx, nodeID, rpc.DefaultClass)
+		if err != nil {
+			log.Warningf(
+				ctx,
+				"Outbox Dial connection error, distributed query will fail: %+v",
+				err,
+			)
+			return err
+		}
 
-	log.VEvent(ctx, 2, "Outbox sending header")
-	// Send header message to establish the remote server (consumer).
-	if err := stream.Send(
-		&execinfrapb.ProducerMessage{Header: &execinfrapb.ProducerHeader{FlowID: flowID, StreamID: streamID}},
-	); err != nil {
-		log.Warningf(
-			ctx,
-			"Outbox Send header error, distributed query will fail: %+v",
-			err,
-		)
+		client := execinfrapb.NewDistSQLClient(conn)
+		stream, err = client.FlowStream(ctx)
+		if err != nil {
+			log.Warningf(
+				ctx,
+				"Outbox FlowStream connection error, distributed query will fail: %+v",
+				err,
+			)
+			return err
+		}
+
+		log.VEvent(ctx, 2, "Outbox sending header")
+		// Send header message to establish the remote server (consumer).
+		if err := stream.Send(
+			&execinfrapb.ProducerMessage{Header: &execinfrapb.ProducerHeader{FlowID: flowID, StreamID: streamID}},
+		); err != nil {
+			log.Warningf(
+				ctx,
+				"Outbox Send header error, distributed query will fail: %+v",
+				err,
+			)
+			return err
+		}
+		return nil
+	}(); err != nil {
+		// error during stream set up.
+		o.close(ctx)
 		return
 	}
 
@@ -324,13 +342,6 @@ func (o *Outbox) runWithStream(
 		}
 	}
 
-	for _, closer := range o.closers {
-		if err := closer.IdempotentClose(ctx); err != nil {
-			if log.V(1) {
-				log.Infof(ctx, "error closing Closer: %v", err)
-			}
-		}
-	}
-
+	o.close(ctx)
 	<-waitCh
 }

